From patchwork Tue Sep  6 19:48:40 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12968065
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 142F1ECAAA1
	for <linux-mm@archiver.kernel.org>; Tue,  6 Sep 2022 19:49:13 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id 86E868D0003; Tue,  6 Sep 2022 15:49:03 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 7DEB68D0001; Tue,  6 Sep 2022 15:49:03 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id 210DE940008; Tue,  6 Sep 2022 15:49:03 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0014.hostedemail.com
 [216.40.44.14])
	by kanga.kvack.org (Postfix) with ESMTP id CC30F6B0080
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 15:49:02 -0400 (EDT)
Received: from smtpin14.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay08.hostedemail.com (Postfix) with ESMTP id A41D3140709
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:02 +0000 (UTC)
X-FDA: 79882698924.14.94F23E2
Received: from mx0a-00069f02.pphosted.com (mx0a-00069f02.pphosted.com
 [205.220.165.32])
	by imf29.hostedemail.com (Postfix) with ESMTP id 2FD0C1200A6
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:02 +0000 (UTC)
Received: from pps.filterd (m0246627.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 286IdAfC020044;
	Tue, 6 Sep 2022 19:48:45 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=/eX6NLsXus/aCvKX6/hTA8FBB3wGYhEify+xnO2FUW8=;
 b=yQISN/wAXbh3mLiSuG1t18OVsJ8AlVVmKHT3hQY1R9yxpNa5ygXrG8p56jFqzvD6yNQN
 soy9nl9yUYdpygviE1JVp8w7GdEGxfef10Ixj42slbDKETDVwtQflrW/XwNsAntWDZK3
 Q1+QPe4MUdaYD3ET8r01k4hz2kiJ72N2KQqC0R94fNHVfxlvpchsE2FjD0CB0jVzUCFr
 m8fPkzCUPxLAa/bqrCr7SUgsqwenbNtJflla81G8JTna0Wh4nNp6sSuotoxTNMcFSCN8
 8U9663DkeJCHgIR1OTtc9xsbWfkBpzkcfVff6NbEvdjjWfAoxdcJhPuOHk7aDQUMBjcX wQ==
Received: from phxpaimrmta02.imrmtpd1.prodappphxaev1.oraclevcn.com
 (phxpaimrmta02.appoci.oracle.com [147.154.114.232])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3jbwh1exjc-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:48:44 +0000
Received: from pps.filterd
 (phxpaimrmta02.imrmtpd1.prodappphxaev1.oraclevcn.com [127.0.0.1])
	by phxpaimrmta02.imrmtpd1.prodappphxaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 286HHdKJ037489;
	Tue, 6 Sep 2022 19:48:44 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2171.outbound.protection.outlook.com [104.47.55.171])
	by phxpaimrmta02.imrmtpd1.prodappphxaev1.oraclevcn.com (PPS) with ESMTPS id
 3jbwc32vns-2
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:48:44 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=OKAv/xRvrCgVX8ur7bIM4etNEiVwp0rdDo0BP8xKr+MJBQ5zCpbFbximNDJZ7bMyVQMhOC3TrYefVlOFXn0VuTjvleW8fHLYYLNOzcg6gV9uUkd/JqYV2pOdCB8GWlMw2RlDpwNQSNeh2cuQPWIQvce1vsocXEOQBEEA9JoQuv+j4Z07I7YsK5hHoljUjgWHvmi+LCcaWtJZAwdCEd15U641+LWtHQTKmRK8rnRIgtiNoD+VDMTcEBGBP6mXcrWklnRkagiG298yVP9zukCIIbyCaRV7NIpqeHRhTy+W20QtWUFkRNG1BrIGXqPh0f6xxeI/uve0EB3ukinllWH7FA==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=/eX6NLsXus/aCvKX6/hTA8FBB3wGYhEify+xnO2FUW8=;
 b=Bf+yZIX/JGQUUwAe4oGgUIfcreGWbFxJGaF3pBF8Srk/QmruRUPahPVcKfUa/TCoXyfUSWnGwChpxBxNL2dVk95wLpW0FXuD4i4DFfmb2soGhyAmUmAOENk5FIRdzSi2iNfgaHm7+LtfRKlIAS3NfkIW3LTFzcQDK4YzDH1o1Q/zO+yp6EMnHI/5w6zf2FBu8RUWaHaNskYdwP0Pb6kZZlZxt8nGJXtcnW7dZleCN1zHb0kruD8V4g9rB6z5KwMczHVgkbj7RQNxjIyiqy1pR3xNmNdqk/GvgohnXnjBGHu57AZY5iejnqbPiMOHUztrbBet7A6UEGkLf48TAZ+WKQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=/eX6NLsXus/aCvKX6/hTA8FBB3wGYhEify+xnO2FUW8=;
 b=Tw17boocl7SYXxSy3kTQDykDGd9wVx/l/9qTSaL7O7cReu2W+BbGOQ+Y61+Fm0aJOUEPyKh9MIQp9/mlVs4oPPyZWlLYqW4URBS2bAZI8vW1tTVREaaKS8mH1IfEWGsuQ200FbDTiphvvXDPvJy0aikVGqsreF/4ol1/IKfJb1M=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by DM4PR10MB6183.namprd10.prod.outlook.com (2603:10b6:8:8b::22) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5588.10; Tue, 6 Sep
 2022 19:48:40 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223%5]) with mapi id 15.20.5588.018; Tue, 6 Sep 2022
 19:48:40 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>
CC: Liam Howlett <liam.howlett@oracle.com>,
        Matthew Wilcox
	<willy@infradead.org>
Subject: [PATCH v14 03/70] radix tree test suite: add
 kmem_cache_set_non_kernel()
Thread-Topic: [PATCH v14 03/70] radix tree test suite: add
 kmem_cache_set_non_kernel()
Thread-Index: AQHYwimpArXaQR/fYkWPJADVIA2tTA==
Date: Tue, 6 Sep 2022 19:48:40 +0000
Message-ID: <20220906194824.2110408-4-Liam.Howlett@oracle.com>
References: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 893b69c3-756b-4766-90bc-08da9040cc44
x-ms-traffictypediagnostic: DM4PR10MB6183:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 NmJ2tgRgMwnaCZbgIFnvOjOyhmWiaVNmmIQxcmREmJLOqN9enVVfHlpvpEA6WdluzoBCN6dYrhG3T1CGn7dq8aosYhDFkAZEU2/smprsLX9yz5YXJVJcprIeiq9HtN3potuqImObc1EEVzClM1REvJ3rVd+H0xcb1dMipyAbBFMnxUpPgvjg63ZrY7VqzU8RY1vc5DcyatpRyQ3+GRW4o8T/xOXlH1O2P9iVYXiBZZfNG0VcsuZrZUbxKDh3l5LeLCkQCUfzjd5pCPcaj3rKdh8j5KgvudDw/mJixGP6rN/Fdjh8kSU4kfAMR1f5vhcsdzqsN52hDY07VkFzq/N8Q9eaPds9IRfb/RyAY1PmirI3rwuJ1xqAhRzH+lfpL+vhSIbIU8zwPIxMlfyipyZmTfm38v8zAEk+qQwPVe8mrrGqG0rUi6tw3rxAHUiwPx22Z6GRhONdWQSPA7PpgUaAPNDC4ZCbE/H0U2MhThTL8BqiRvSj2wTGsusnG5S+qBIWBjdMyh4oBTnBc7Wue7ZtyEKepxMJbpQTIiCZ7RtkPZCCcgl/icxq6+WXnnWh48MdkN5hZwwtk53WHwxSJm36ZhC4WRlKCS54uxCIEB7aAwD2549qKlmmPPFieS4TTa9D+e15/Z+afM7YtZ6nNJR4g+EllxSSGlsTRlICvdgxEq1boCc44b8MKdKL7dDlhUTqWT+7y3HCLXIwTHf/pLzMMntD0LLB7UTsQr5GhlnLe2ExgC6qlyFFUuPpFYE9DWjB93BnMLTRABkxqh5AEoTZhg==
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(39860400002)(366004)(376002)(396003)(346002)(136003)(66556008)(38100700002)(122000001)(8936002)(5660300002)(76116006)(66946007)(66446008)(66476007)(2906002)(83380400001)(8676002)(91956017)(36756003)(2616005)(186003)(86362001)(1076003)(110136005)(44832011)(71200400001)(38070700005)(54906003)(26005)(64756008)(41300700001)(6506007)(6512007)(6486002)(316002)(478600001)(4326008);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?ZaLwD4MXZF8Eo5gHiegReT0?=
	=?iso-8859-1?q?B+BOtNlq6zm0ZarnuCXzAuTeAYrIKj9qH0GCspqaKJ8QOcy5zUSalffE/eUM?=
	=?iso-8859-1?q?NfQpCgqLOKw0nX9DbgNW4efkvZXFola/0qik3ihSKsk2YRF6TFPMezn2uXhw?=
	=?iso-8859-1?q?/5Q3kphkYF+fvgIjvJbj0gFv69TnCWxU3BvhRqecOced48MN4zRbqYxnNyTP?=
	=?iso-8859-1?q?6A9cPPNmfMPBF4f7F+6vc8WM4A2OPqZBGwWRw5BtyEYheN/ROS3EKYFbeqi/?=
	=?iso-8859-1?q?sEwi6N5mWTNylrbi4PQVeL2Lom44da2Xn7icBpKDWbCUEVPaHPHtaavFs2tR?=
	=?iso-8859-1?q?Xzy1hvWTLuf5TvSr6u0u+w8jWwxHvvgzPoeK509o6YkgbnA26UxG+LPXxrAk?=
	=?iso-8859-1?q?flNQEHALnjYTUBgcrRuqCwEd+JoD99jjhYByIYSfSXVtGZi1ipims72VVo0H?=
	=?iso-8859-1?q?HOZkC3dsrWHVDBDBPzWVbx6v2YieTlthLdH7NL+IaTWc0Tv+NG2idyFps+B9?=
	=?iso-8859-1?q?GpLEgiZK6p3cFaYDrLItgYJwxE9yzU0mb11/GeeJ0ag+bh9YQRkEQK7jBU7I?=
	=?iso-8859-1?q?5Y1NKsDPliQUAbP9krt1DbxlGZ+c3amCWAvgavvS1Xz8jaeuA/UczxE5XiXB?=
	=?iso-8859-1?q?5lLK5w8juAKSPJwyiHTv9nQWaU0E3qcFPw+XhbMRAFPFQsGQqTO9G72vkczV?=
	=?iso-8859-1?q?LSYwCznEUREPm1gE7OZTD5jb/8ObQuQRROnOCnMFAJwIApYUCvapFe3rIDbN?=
	=?iso-8859-1?q?VYW7/NpEIOj3JYXLHjBcNiXfvi1t9U87uUEG9xFeX+8XA8A2vppAPp03q+9U?=
	=?iso-8859-1?q?ih54YmaVCb2sKQlNfeXWBc0ANLUO/sMtlZzoAB2Wbo/YugvSaVFbJg9F0b2r?=
	=?iso-8859-1?q?FxjdqHDGv+oIC0jTBw7PhwzbgxhfPxX+gEQE3Ui2jXDsw+Yv2VZ01UO0aJxQ?=
	=?iso-8859-1?q?IkS1hoH4jdHIBj63miSGcaLepbDyS32IbRs8B01tLzeiM1Cpzq2KibxBfHSt?=
	=?iso-8859-1?q?Se8cH4iT7ojDlYTZAm5uqDjlVA3MOBxlovhMFYhbiSX3CuTTSh9C88nPqmae?=
	=?iso-8859-1?q?zolvL9WKXEBUWxSURFD7KYiDPCg66Ypmi3vOjRPX5bTWPOTDa6x867o1X9D8?=
	=?iso-8859-1?q?TVqWrPX0+Br4fBCHwMfAnyOSZqi+bQ9KqKZdyZ/tsgEr06DTbQeHg42Dq2rz?=
	=?iso-8859-1?q?a7fvtT33YSk6J7twMvVOQJa2w2xtWFjjRpLcmBPT1TGjDwbYITEaiu4+/tUH?=
	=?iso-8859-1?q?Iq+gIBV5YDwHe0XP6PcvA0HhAb3yEGkPf2g3uaJT792A8R6rAOeTGTwBx0VM?=
	=?iso-8859-1?q?xw4dCJB0bkCuVnoJBpVQzyfj/R13B9DDAzjBCjUi2rSWEyAV3BmO2NAjmnP/?=
	=?iso-8859-1?q?l0El4IrqsactaAUIVj9cLWDIuPHqAemqRucC4luxR2tNHSl+IReX1ryWVJSi?=
	=?iso-8859-1?q?YtR4HayS+Vsz3qFrT3sOsoogOF/mW6Dv5caHoQPA4LRnJCcnp0MVySPjIXIj?=
	=?iso-8859-1?q?yJwARv84Mcw+Ut3qNFDuk82eOAdC/GKIy4r3DUcZsRrJ8UM0CZwgDrhb8ymm?=
	=?iso-8859-1?q?FYDxozd3wHwRDhiIylrCfM4EvAvH0oP7R42zZ80AsI5Th92Iz5EyfIqRZ3JW?=
	=?iso-8859-1?q?e0YT4rQyWWX2/CSuxpkcCk1tLTmFnNyHfiBoBJg=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 893b69c3-756b-4766-90bc-08da9040cc44
X-MS-Exchange-CrossTenant-originalarrivaltime: 06 Sep 2022 19:48:40.2573
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 AFxC3HKMmeOkJUE+5TN5MvXV4JjFGCMhPY7Kj8Y5qoGj+uBvgc0SW9Yprs4ml2zvbscSjXoCdmnfBLp/F8L9HQ==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: DM4PR10MB6183
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.895,Hydra:6.0.528,FMLib:17.11.122.1
 definitions=2022-09-06_09,2022-09-06_02,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 bulkscore=0
 suspectscore=0 phishscore=0
 mlxlogscore=999 spamscore=0 adultscore=0 mlxscore=0 malwarescore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2207270000
 definitions=main-2209060091
X-Proofpoint-GUID: 7XPWf0dXsjNFRhpJdmvQb3exR6I1QhDZ
X-Proofpoint-ORIG-GUID: 7XPWf0dXsjNFRhpJdmvQb3exR6I1QhDZ
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1662493742; a=rsa-sha256;
	cv=pass;
	b=iAj9m15ylgwAC7Bh0X+BF+jQyV7xyOcq93nQNrTA1rblIzsyNOT9ZFpB+Dq/QO3o5zU9tL
	A995urmGw1ivr3nuyQkRJAXoLARXGxOq0XJWAAHHXphFvQYZa60GjUha8U5l/zOghVzz5o
	Fztl0SJsYeAJHfqxe5zJeJ5knw6rmuU=
ARC-Authentication-Results: i=2;
	imf29.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b="yQISN/wA";
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=Tw17booc;
	spf=pass (imf29.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1662493742;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=/eX6NLsXus/aCvKX6/hTA8FBB3wGYhEify+xnO2FUW8=;
	b=PDHNZUcQ7TBq08lBwL0a53egIjg+ECXsmx5R87oIs7HDOw7jZAdE7j8pvR4y99AbLkGXe+
	PKqWw5D+RdWIyxqRO6wWPs+jc1naDcTq958JhipFxFSQ59dA6mDmFWmy7Mqq8Rt9Tjg7y3
	kgxg/FwZD7nhlcCA3oXsSWSViUb3K5E=
Authentication-Results: imf29.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b="yQISN/wA";
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=Tw17booc;
	spf=pass (imf29.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
X-Rspam-User: 
X-Rspamd-Server: rspam06
X-Stat-Signature: 98wa5h8naaigh64nwkhgkd5egj6kh4md
X-Rspamd-Queue-Id: 2FD0C1200A6
X-HE-Tag: 1662493741-178271
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Liam R. Howlett" <Liam.Howlett@Oracle.com>

kmem_cache_set_non_kernel() is a mechanism to allow a certain number of
kmem_cache_alloc requests to succeed even when GFP_KERNEL is not set in
the flags.  This functionality allows for testing different paths though
the code.

Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
---
 tools/testing/radix-tree/linux.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/tools/testing/radix-tree/linux.c b/tools/testing/radix-tree/linux.c
index d5c1bcba86fe..277aa8b70abc 100644
--- a/tools/testing/radix-tree/linux.c
+++ b/tools/testing/radix-tree/linux.c
@@ -23,15 +23,26 @@ struct kmem_cache {
 	int nr_objs;
 	void *objs;
 	void (*ctor)(void *);
+	unsigned int non_kernel;
 };
 
+void kmem_cache_set_non_kernel(struct kmem_cache *cachep, unsigned int val)
+{
+	cachep->non_kernel = val;
+}
+
 void *kmem_cache_alloc_lru(struct kmem_cache *cachep, struct list_lru *lru,
 		int gfp)
+
 {
 	void *p;
 
-	if (!(gfp & __GFP_DIRECT_RECLAIM))
-		return NULL;
+	if (!(gfp & __GFP_DIRECT_RECLAIM)) {
+		if (!cachep->non_kernel)
+			return NULL;
+
+		cachep->non_kernel--;
+	}
 
 	pthread_mutex_lock(&cachep->lock);
 	if (cachep->nr_objs) {
@@ -90,5 +101,6 @@ kmem_cache_create(const char *name, unsigned int size, unsigned int align,
 	ret->nr_objs = 0;
 	ret->objs = NULL;
 	ret->ctor = ctor;
+	ret->non_kernel = 0;
 	return ret;
 }

